cdap-data-fabric/src/main/java/co/cask/cdap/data2/registry/DefaultUsageRegistry.java

```diff
@@ -0,0 +1,317 @@
+   */
+  public void register(Id user, Id.Stream streamId) {
+    if (user instanceof Id.Program) {
+      register((Id.Program) user, streamId);
+    }
+  }
+
+  /**
+   * Registers usage of a stream by multiple ids.
+   *
+   * @param users     the users of the stream
+   * @param datasetId the stream
+   */
+  public void registerAll(final Iterable<? extends Id> users, final Id.DatasetInstance datasetId) {
+    for (Id user : users) {
+      register(user, datasetId);
+    }
+  }
+
+  /**
+   * Registers usage of a dataset by multiple ids.
+   *
+   * @param user      the user of the dataset
+   * @param datasetId the dataset
+   */
+  public void register(Id user, Id.DatasetInstance datasetId) {
+    if (user instanceof Id.Program) {
+      register((Id.Program) user, datasetId);
+    }
+  }
+
+  /**
+   * Registers usage of a dataset by a program.
+   *
+   * @param programId         program
+   * @param datasetInstanceId dataset
+   */
+  public void register(final Id.Program programId, final Id.DatasetInstance datasetInstanceId) {
+    usageCache.getUnchecked(new DatasetUsageKey(datasetInstanceId, programId));
+  }
+
+  /**
+   * Internal method to register usage of a dataset by a program, called from the cache loader.
+   *
+   * @param programId program
+   * @param datasetInstanceId  dataset
+   */
+  private void doRegister(final Id.Program programId, final Id.DatasetInstance datasetInstanceId) {
+    execute(new TransactionExecutor.Procedure<UsageDataset>() {
+      @Override
+      public void apply(UsageDataset usageDataset) throws Exception {
+        usageDataset.register(programId, datasetInstanceId);
+      }
+    });
+  }
+
+  /**
+   * Registers usage of a stream by a program.
+   *
+   * @param programId program
+   * @param streamId  stream
+   */
+  public void register(final Id.Program programId, final Id.Stream streamId) {
+    execute(new TransactionExecutor.Procedure<UsageDataset>() {
+      @Override
+      public void apply(UsageDataset usageDataset) throws Exception {
+        usageDataset.register(programId, streamId);
+      }
+    });
+  }
+
+  /**
+   * Unregisters all usage information of an application.
+   *
+   * @param applicationId application
+   */
+  public void unregister(final Id.Application applicationId) {
+    execute(new TransactionExecutor.Procedure<UsageDataset>() {
+      @Override
+      public void apply(UsageDataset usageDataset) throws Exception {
+        usageDataset.unregister(applicationId);
+      }
+    });
+
+    // we must invalidate the cache for all programs of this application. Because if, for example, an
+    // application is deleted, its usage is removed from the registry. If it is redeployed later, we
+    // must register its usage again. That would not happen if the cache still holds these entries.
+    for (DatasetUsageKey key : usageCache.asMap().keySet()) {
+      if (applicationId.equals(key.getOwner().getApplication())) {
+        usageCache.invalidate(key);
```

Comment:
Shall it be done in the `apply` method above? In case there is exception for some of the `unregister` call, the cache will still in sync with the dataset.